pkg/api/types.go

```diff
@@ -488,3 +501,73 @@ type SuspendAgent struct {
 type ResumeAgent struct {
 	ExpectedID uuid.UUID `json:"expectedID"`
 }
+
+////////////////////////////////////
+// Controller <-> Runner Messages //
+////////////////////////////////////
+
+// VCPUChange is used to notify runner that it had some changes in its CPUs
+// runner uses this info to adjust qemu cgroup
+type VCPUChange struct {
+	VCPUs resource.Quantity
+}
+
+// VCPUCgroup is used in runner to reply to controller
+// it represents the vCPU usage as controlled by cgroup
+type VCPUCgroup struct {
+	VCPUs resource.Quantity
+}
+
+// MilliCPU is a special type to represent vCPUs * 1000
+// e.g. 2 vCPU is 2000, 0.25 is 250
+type MilliCPU uint32
+
+// MilliCPUFromResourceQuantity converts resource.Qunatity into MilliCPU
+func MilliCPUFromResourceQuantity(r resource.Quantity) MilliCPU {
+	return MilliCPU(r.MilliValue())
+}
+
+// ToResourceQuantity converts a MilliCPU to resource.Qunatity
+// this is useful for formatting/serialization
+func (m *MilliCPU) ToResourceQuantity() resource.Quantity {
+	return *resource.NewMilliQuantity(int64(*m), resource.BinarySI)
+}
+
+// this is used to parse scheduler config and communication between components
+// we used resource.Quantity as underlying transport format for MilliCPU
+func (m *MilliCPU) UnmarshalJSON(data []byte) error {
+	var quantity resource.Quantity
+	err := json.Unmarshal(data, &quantity)
+	if err != nil {
+		return err
+	}
+
+	*m = MilliCPUFromResourceQuantity(quantity)
+	return nil
+}
+
+func (m *MilliCPU) MarshalJSON() ([]byte, error) {
+	return json.Marshal(m.ToResourceQuantity())
+}
+
+func (m MilliCPU) Format(state fmt.State, verb rune) {
+	switch {
+	case verb == 'v' && state.Flag('#'):
+		state.Write([]byte(fmt.Sprintf("%v", uint32(m))))
+	default:
+		quantity := m.ToResourceQuantity()
+		state.Write([]byte(fmt.Sprintf("%v", quantity.AsApproximateFloat64())))
+	}
+}
+
+// this a similar version type for controller <-> runner communications
+// see PluginProtoVersion comment for details
+type RunnerProtoVersion uint32
+
+const (
+	RunnerProtoV1_0 RunnerProtoVersion = iota + 1
+)
+
+func (v RunnerProtoVersion) SupportsCgroup() bool {
```

Comment:
Name here maybe could be improved? I'd probably say `SupportsCgroupFractionalCPU`, but I don't know. I'm thinking along the lines of: "Can this be more clear to someone who doesn't know the details runner's implementation?".

Thoughts?